src/keys.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::{read_dir, DirEntry, OpenOptions};
+use std::io::{self, Read, Write};
+use std::os::unix::ffi::OsStringExt;
+use std::os::unix::fs::PermissionsExt;
+use std::path::{Path, PathBuf};
+
+use error_chain::{bail, ChainedError};
+
+use crate::errors::{Result, ResultExt};
+
+const KEYS_SUBDIR: &str = "authorized_keys.d";
+
+pub(crate) fn read_keys(ssh_dir: &Path, out: &mut dyn Write, err: &mut dyn Write) -> Result<()> {
+    // check permissions on paths leading to KEYS_SUBDIR
+    ensure_safe_dir_mode(ssh_dir)?;
```

Comment:
Hmm, does this mean we error out if `~/.ssh` doesn't exist? Why is that an error instead of a no-op? (And similarly below for the authorized keys dir).

Reply:
I thought it was harmless, but it turns out that sshd does log non-zero exit status from an `AuthorizedKeysCommand`, so you're right.